apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -2491,12 +2544,110 @@ is_channel_apt_for_queries(_) ->
 
 -spec request_context(query_opts()) -> request_context().
 request_context(QueryOpts) ->
-    maps:with([is_fallback, fallback_actions], QueryOpts).
+    maps:with(
+        [
+            ?is_fallback,
+            ?fallback_actions,
+            ?reply_to
+        ],
+        QueryOpts
+    ).
 
 -spec result_context([queue_query()]) -> query_result_context().
 result_context(Queries) ->
     #{?queries => Queries}.
 
+%% Should be called for individual queries, not batches.
+%% `batch_reply_caller_defer_metrics' handles splitting a batch into individual result
+%% contexts.
+-spec maybe_trigger_fallback_actions(id(), query_result_context()) -> ok.
+maybe_trigger_fallback_actions(_Id, #{
+    ?queries := [
+        ?QUERY(_, _, _, _, #{?is_fallback := true}, _)
+    ]
+}) ->
+    ok;
+maybe_trigger_fallback_actions(Id, #{
+    ?queries := [
+        ?QUERY(
+            _From,
+            {Id, #{} = Req},
+            _HasBeenSent,
+            _ExpireAt,
+            #{?fallback_actions := [_ | _] = FallbackActions} = RequestContext,
+            TraceCtx
+        )
+    ]
+}) ->
+    QueryOpts0 = maps:with([?reply_to], RequestContext),
+    QueryOpts = QueryOpts0#{
+        ?is_fallback => true,
+        query_mode => async,
```

Comment:
maybe add comment to doc why query mode is always async here.

also a question: is user allowed to create the action with sync query?

Reply:
yes, the resource query mode is separate from the request query kind

Reply:
:checkered_flag: